File: src/checks/technical/robots-txt.ts
import type { Check, CheckContext, CheckResult } from "../../types.js";

const ID = "robots-txt";
const NAME = "robots.txt";
const CATEGORY = "technical" as const;

export const robotsTxtCheck: Check = {
  id: ID,
  name: NAME,
  category: CATEGORY,
  run(ctx: CheckContext): CheckResult {
    const base = { id: ID, name: NAME, category: CATEGORY };
    const robots = ctx.resources.robotsTxt;

    if (!robots.ok) {
      return {
        ...base,
        status: "FAIL",
        message: `robots.txt was not reachable at ${robots.url}${robots.status ? ` (HTTP ${robots.status})` : robots.error ? ` (${robots.error})` : ""}.`,
        fix: "Add a robots.txt file at your site root, even a permissive one, so crawlers and agents have explicit directives.",
      };
    }

    const body = robots.body ?? "";
    if (!/user-agent\s*:/i.test(body)) {
      return {
        ...base,
        status: "WARN",
        message: "robots.txt is reachable but contains no User-agent directive; it may not be a valid robots.txt file.",
        fix: "Ensure robots.txt contains at least one User-agent block.",
      };
    }

    return {
      ...base,
      status: "PASS",
      message: `robots.txt is reachable at ${robots.url} and contains User-agent directives.`,
    };
  },
};

File: python/src/seofleet/checks/technical/robots_txt.py
"""Ported from src/checks/technical/robots-txt.ts."""
from __future__ import annotations

import re

from ...types import Check, CheckContext, CheckResult

_ID = "robots-txt"
_NAME = "robots.txt"
_CATEGORY = "technical"
_USER_AGENT_RE = re.compile(r"user-agent\s*:", re.IGNORECASE)


def _run(ctx: CheckContext) -> CheckResult:
    robots = ctx.resources.robots_txt

    if not robots.ok:
        detail = f" (HTTP {robots.status})" if robots.status else (f" ({robots.error})" if robots.error else "")
        return CheckResult(
            _ID, _NAME, _CATEGORY, "FAIL",
            f"robots.txt was not reachable at {robots.url}{detail}.",
            "Add a robots.txt file at your site root, even a permissive one, so crawlers and agents have explicit directives.",
        )

    body = robots.body or ""
    if not _USER_AGENT_RE.search(body):
        return CheckResult(
            _ID, _NAME, _CATEGORY, "WARN",
            "robots.txt is reachable but contains no User-agent directive; it may not be a valid robots.txt file.",
            "Ensure robots.txt contains at least one User-agent block.",
        )

    return CheckResult(
        _ID, _NAME, _CATEGORY, "PASS",
        f"robots.txt is reachable at {robots.url} and contains User-agent directives.",
    )


robots_txt_check = Check(id=_ID, name=_NAME, category=_CATEGORY, run=_run)
